pkg/trie/triedb/proof/generate.go

```diff
@@ -0,0 +1,455 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package proof
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/trie"
+	nibbles "github.com/ChainSafe/gossamer/pkg/trie/codec"
+	"github.com/ChainSafe/gossamer/pkg/trie/db"
+	"github.com/ChainSafe/gossamer/pkg/trie/triedb"
+	"github.com/ChainSafe/gossamer/pkg/trie/triedb/codec"
+	"github.com/gammazero/deque"
+	"golang.org/x/exp/slices"
+)
+
+type MerkleProof [][]byte
+
+type nodeHandle interface {
+	isNodeHandle()
+}
+
+type (
+	nodeHandleHash struct {
+		hash []byte
+	}
+
+	nodeHandleInline struct {
+		data []byte
+	}
+)
+
+func (nodeHandleHash) isNodeHandle()   {}
+func (nodeHandleInline) isNodeHandle() {}
+
+type step interface {
```

Comment:
nit/maybe `proofStep` instead

Reply:
I renamed it in my last [PR](https://github.com/ChainSafe/gossamer/pull/4091/files#diff-29ba1e2810f9ece35c52a7fd9e3de4a554eda5cb9787e64362a70866e0112bc8)